File: typescript/strain/strain.ts
const keep = <T>(array: T[], predicate: (item: T) => boolean): T[] =>
    filter(array, predicate)

const discard = <T>(array: T[], predicate: (item: T) => boolean): T[] =>
    filter(array, (item) => !predicate(item))

const filter = <T>(array: T[], predicate: (item: T) => boolean): T[] =>
    filterRec(array, [], predicate)

const filterRec = <T>(array: T[], acc: T[], predicate: (item: T) => boolean): T[] => {
    if (array.length === 0) {
        return acc
    }
    const [head, ...tail] = array
    const newAcc = predicate(head) ? [...acc, head] : acc
    return filterRec(tail, newAcc, predicate)
}

export {keep, discard}

File: haskell/strain/src/Strain.hs
module Strain
  ( keep
  , discard
  ) where

discard :: (a -> Bool) -> [a] -> [a]
discard p = keep (not . p)

keep :: (a -> Bool) -> [a] -> [a]
keep predicate = foldr (appendIf predicate) []

appendIf :: (a -> Bool) -> a -> [a] -> [a]
appendIf predicate current acc
  | predicate current = current : acc
  | otherwise = acc